ansible_collections/arista/avd/roles/eos_designs/python_modules/pool_manager/avdpoolmanager.py

```diff
@@ -0,0 +1,168 @@
+            id_file.touch(mode=0o664)
+
+        # Read from file or assign an empty data model.
+        id_data = safe_load(id_file.read_text(encoding="UTF-8")) or {"id_pools": []}
+
+        if not isinstance(id_data, dict) or not isinstance(id_data.get("id_pools"), list):
+            raise AristaAvdError(f"Invalid ID pool manager data when reading {id_file}. Expecting {'id_pools': []}")
+
+        # Since we assigned a new ID, we have to mark the id pool as changed to have it saved once the build is done.
+        self.id_files.add(id_file)
+
+        return id_data
+
+    def get_id(self, shared_utils: SharedUtils) -> int | None:
+        """
+        Returns the ID of one node if found in the pool.
+        Otherwise a new entry is inserted into the pool and the ID is returned.
+
+        The pools are dynamically built and matched on the following device data:
+        - Fabric Name (from 'shared_utils.hostvars.fabric_name')
+        - DC Name (from `shared_utils.dc_name`)
+        - POD Name (from `shared_utils.pod_name`)
+        - Node type (from `shared_utils.type`)
+
+        Each pool will assign first available ID starting from 1.
+
+        Args:
+            shared_utils: Instance of SharedUtils initialized for one device.
+
+        Returns:
+            Node ID
+        """
+        fabric_name = get(shared_utils.hostvars, "fabric_name", required=True)
+        default_id_file = f"{self.output_dir}/data/{fabric_name}-ids.yml"
+        id_file = Path(get(shared_utils.hostvars, "fabric_numbering.id.pools_file", default=default_id_file))
+        id_data = self.id_pool_data(id_file)
+
+        id_pool = None
+        for id_data_id_pool in id_data["id_pools"]:
+            if (
+                id_data_id_pool.get("fabric_name") == fabric_name
+                and id_data_id_pool.get("dc_name") == shared_utils.dc_name
+                and id_data_id_pool.get("pod_name") == shared_utils.pod_name
+                and id_data_id_pool.get("type") == shared_utils.type
+            ):
+                id_pool = id_data_id_pool
+                break
+
+        if not id_pool:
+            # Create a new dict, add to the list of pools.
+            id_pool = {
+                "fabric_name": fabric_name,
+                "dc_name": shared_utils.dc_name,
+                "pod_name": shared_utils.pod_name,
+                "type": shared_utils.type,
+                "id_assignments": [],
+            }
+            id_data["id_pools"].append(id_pool)
+
+        if id_assignment := get_item(id_pool["id_assignments"], "hostname", shared_utils.hostname):
+            # Found an ID, so return quickly.
+            id_assignment["accessed"] = True
+            return id_assignment["id"]
+
+        pool_sorted_on_id = sorted(id_pool["id_assignments"], key=lambda x: x.get("id"))
+        first_available_id = next((index for index, id_assignment in enumerate(pool_sorted_on_id + [{"id": 0}], 1) if id_assignment["id"] != index))
```

Comment:
To catch the empty list. This is the most effective method, but we need to change this algorithm since it does not handle duplicate entries. So if a user edits, and create duplicates, we cannot find the correct next-available.

Reply:
I have changed the algorithm to something that is simpler to read (I hope) and also handles duplicate entries.